dartagnan/src/main/java/com/dat3m/dartagnan/expression/ExpressionFactory.java

```diff
@@ -21,163 +26,207 @@ public static ExpressionFactory getInstance() {
     // Boolean
 
     public BConst makeTrue() {
-        return BConst.TRUE;
+        return trueConstant;
     }
 
     public BConst makeFalse() {
-        return BConst.FALSE;
+        return falseConstant;
     }
 
     public BConst makeValue(boolean value) {
         return value ? makeTrue() : makeFalse();
     }
 
-    public BExpr makeNot(Expression operand) {
+    public Expression makeNot(Expression operand) {
         return makeUnary(BOpUn.NOT, operand);
     }
 
-    public BExpr makeUnary(BOpUn operator, Expression operand) {
-        return new BExprUn(operator, operand);
+    public Expression makeUnary(BOpUn operator, Expression operand) {
+        return new BExprUn(types.getBooleanType(), operator, operand);
     }
 
-    public BExpr makeAnd(Expression leftOperand, Expression rightOperand) {
+    public Expression makeAnd(Expression leftOperand, Expression rightOperand) {
         return makeBinary(leftOperand, BOpBin.AND, rightOperand);
     }
 
-    public BExpr makeOr(Expression leftOperand, Expression rightOperand) {
+    public Expression makeOr(Expression leftOperand, Expression rightOperand) {
         return makeBinary(leftOperand, BOpBin.OR, rightOperand);
     }
 
-    public BExpr makeBinary(Expression leftOperand, BOpBin operator, Expression rightOperand) {
-        return new BExprBin(leftOperand, operator, rightOperand);
+    public Expression makeBinary(Expression leftOperand, BOpBin operator, Expression rightOperand) {
+        return new BExprBin(booleanType, leftOperand, operator, rightOperand);
     }
 
     public IValue makeZero(IntegerType type) {
         return makeValue(BigInteger.ZERO, type);
     }
 
+    @Deprecated
+    public IValue makeZero(Type type) {
+        if (type instanceof IntegerType t) {
+            return makeZero(t);
+        }
+        throw new IllegalArgumentException("Non-integer type " + type);
```

Comment:
You should use `Preconditions.checkArgument` instead.
This is also true for the other methods like that raise `IllegalArgumentExceptions` in a similar way.